### PR TITLE
Fix 2055: multiple if statements as list comprehension

### LINKED ISSUE #2055: Idempotency and correctness problem when using multiple if statements as list comprehension
Issue created from [fantomas-online](https://fsprojects.github.io/fantomas-tools/#/fantomas/preview?data=N4KABGBEDGD2AmBTSAuKAbRAXMB3ATrAHYDmYAvGANoA6%2BNRETTAlgGZhsCG6AzomCwALRIwCMdBs2btB%2BAK4DhosACZJjJgF1IAGnBR%2BWLC1K9U1AxFDSoAEiwBPAA7I0kU1j1WmkOwDceRQsqH2YABn1bCEgASSIkIiwAZRYAL2Qo6IAWMLAtHwBfLOs8vydXCw8k71s-QPRgtFDosDES6UgAWS4ADwAZU0R%2B0RJhWuixVXC8guZinxtOhxc3KFF4AHk2QaJEZKdMCeZ6oLWW6NUOk4BRBO3dzLyY9DZIWaKOpZOVyvcAI1gsHQx18ATOIWeYAAzNdfPF%2BPgsAAxUw8AByiFw6CGoOYWAUiA%2B8y%2BZV%2Ba0ggOBePsDSalla2ThMWSiAAtiw4OhiABBLB3LY7XHMzg8fjEpgLZjfMEVClUkFw06Nc5QgCsIsgyWcXGgiAAQog2LB8IgAApcfBcNnYRD4GkQAmKCUQKVMGUxclVBU05X0i62ABsmu1uoNRpNw1guDt0C4-Hi-lgcZMxAdcmdtjmktJdS9AKBirJdNVrQA7CGdXrDcbTQBVZyufBxhNEJMplhpkXcPhErOfRZkuXewu%2B8EqyGtAAclbDNcjAGF0PHeAviLwndAsCb0z3xf2SYO88OC9SleP-VCAJyz6sR01ddn-O27sV96TZ1255YnqA%2B88lpOkyRFCWpVuGtaIGu3JEK%2BvYumAbqlMeqwjmexYQs0UJiO0oGhnqPJsFgdprmybJcOmTrvswn6Id%2BPy-pSo4AZhDKTFceHgfOpqshyXJdlCe7UdoA7SkOqGnkWeaAVhrRiLCnFhoRxH4LxnLAgJrRUQhSFgB69iMf%2BGETrJkxMqB8SJFgmxEAAKvgjgAOosOM3Zvjp9GyhJf7McZl5yRqikEYQ8gJAAIogOIciplGEh5R4-t51ReCxJlsbYYjBqBPS9LEbC2SIRA3L2yRCCaWDOfArlQtkMwHjmCUMUlnhjjJ6XSGIFbZX08RsCwvSbE2XDbvgNy9M4pq8LwnawSKap1R%2BonuuJfxQC1qX%2BZMM7db0ABKiBwPg8CVdVjILTRS3IYlq3Ja1rEBh1N47fth3wOi8hss%2B%2BDbLExFsuYIpiPFYkoTdbLyOgJg4nsyImuRxh2rZqEbaWlwga0kAvSa8BdBDUNDLD%2BDwzFmrQEIlq6ipAD6uAsFVQjvPVX6NV5N3rX5qO2KouEYzlPL4FajibPggwbidDMirVwPLaDFLs9J91QqoHG830-OC8LotYO9n12j9f0A9h0tXU1YN4yw0OIITxOI8jHNAVzCkY%2BrXBCyLLAbrjkMWwTcPDSToFkxTW52jTdPjMbekrXLNQow70iqOZqu9AAamc%2BqmPApgkOL6ZTudImHiD10xyl9umVzgXJ8ioVbjNGcJNnueSwXEC0bp%2BnlM1sfl%2B1zCqFlydhbAWAAOLYGNE2IFNM3N%2Bqrf5JdUey2hUnLG1D3911vPm5b%2BrctAADW%2BpWkf2C8DZyTWlBwIfbNgnuUzdEs56hm%2BQraWb0wqjbRjmLYkMQ0WAYyiCRq4CKfUiAuRmjyBIj5db4ENq0ISkdO75h8uhD%2Bm0uZPQxgAaUQIgZweUCqiHiFfG0jw4L7kWkXGWJcqjy3XorVo0J0bRG6H0YqHJeBCDnmdVB0dV53U-lCaEPMOGpFIJgfmJAPqiAqogf4XQEBPGQY-WhDVi6m3lO-ZhojWEqw4TyHEJAiA1yIHXYgqQzHDXkKaWysBLIKOGjNahwk25LzQW-TB%2BjsHSGhE7YxkM7REFcf4KMpB4FfQgaYaB653GCJXpJER-jmDQiThwr2JhHj9GtP8eAXAlywGmqQf%2BltElPw7kIlJccK4BKrhwsKHsuD-EwNwj2QhkiOCSH0SpmjmbaNZro3xPwN5iMHhwghRDnFJHiCfLgliJYP3glUzyr8kpGSwZzAJ28OH70WYfR4vB%2BawFCm9aexEca7yGJPSa00Ekim0msl%2BBlNl6LGSw6I0Jf77MtNxRAzTeDQHwCwDkYSrl1igcQCK0BlxWlTPfdRqyBnPyGRsm6Wy-E7PSbgyRBJORYBUUgfpF15gGC0FkDwvBkTTQsCgwoQA)

#### Original code

```fsharp
let wrong = [
        if false then 1
        if true then 2
    ]

```

When executed:
```
> let wrong = [
        if false then 1
        if true then 2
    ];;
val wrong: int list = [2]
```

#### Formatted code

```fsharp
let wrong = [ if false then 1; if true then 2 ]

```

When executed:
```
> let wrong = [ if false then 1; if true then 2 ];;
val wrong: int list = []
```
            

#### Reformatted code

```fsharp
let wrong =
    [ if false then
          1
          if true then 2 ]

```

When executed:
```
> let wrong =
    [ if false then
          1
          if true then 2 ];;
val wrong: int list = []
```
      
#### Problem description

* Fantomas broke the correctness of my code when first formatting the code.
* Fantomas was not able to produce the same code after reformatting the result.

#### Extra information

- [x] The formatted result breaks by code.
- [ ] The formatted result gives compiler warnings.
- [ ] I or my company would be willing to help fix this.

#### Options

Fantomas master branch at 1/1/1990

Default Fantomas configuration

### PR DESCRIPTION
**Pull request by dawedawe**
Fixes #2055

Not totally sure, if this is the fix you had in mind, but could be a start.

### CODE REVIEW

File: src/Fantomas/CodePrinter.fs [hunk 1]
```diff
@@ -1339,7 +1339,7 @@ and genExpr astContext synExpr ctx =
 
             fun ctx ->
                 if List.exists isIfThenElseWithYieldReturn xs
-                   || List.forall isSynExprLambda xs then
+                   || List.forall (fun x -> isSynExprLambda x || isIfThenElse x) xs then
```
Comment (nojaf) [inline src/Fantomas/CodePrinter.fs:1342]: As `isSynExprLambda` is only used here, I would change that function to also take `SynExpr.IfThenElse` and rename it to something like `isDangerousExprInSingleLineArrayOrList` (or some better name 😅)

**Review by nojaf — CHANGES_REQUESTED** [review]
Yep, this is roughly what I had in mind.
Thanks again David!

File: src/Fantomas/CodePrinter.fs (hunk 1 above)
Comment (dawedawe) [reply src/Fantomas/CodePrinter.fs]: Ah yes, I thought about that, too. But somehow expected that separate functions would be preferred.
Anyway, naming is hard...

**Review by nojaf — APPROVED** [review]
Many thanks!

File: src/Fantomas/CodePrinter.fs (hunk 1 above)
Comment (nojaf) [reply src/Fantomas/CodePrinter.fs]: Thanks!

**Comment by OkkeHendriks** [comment]
Thanks @dawedawe !

**Comment by nojaf** [comment]
@OkkeHendriks do you want a release for this fix?

**Comment by OkkeHendriks** [comment]
Always nice :) but there is no direct hurry! 

I worked around it, for now, as follows:

``` F#
let correct = seq {
  if false then 1
  if true then 2
}
```

**Pull request merged**